### DISCUSSION TITLE
Not all files and folders removed by Joomla\CMS\Installer\Adapter\FileAdapter

### DISCUSSION THREAD

**Discussion by pjdevries** [joomla/joomla-cms#38410] [Q&A]
It seems that `Joomla\CMS\Installer\Adapter\FileAdapter` does not delete all files and folders that were installed. This piece of code, starting at line 292, is the culprit:

```php
	// Delete any folders that don't have any content in them.
	foreach ($folderList as $folder)
	{
		$files = Folder::files($folder);

		if ($files !== false && !\count($files))
		{
			Folder::delete($folder);
		}
	}
```

Why aren't all folders deleted recursively? Is this simply a mistake or intentional behavior? And if intentional, what should an installer manifest look like, that installs several files and sub folders, containing files as well, in for instance `layouts\com_foo`?

**Comment by richard67** [comment]
`Folder::delete` **_does_** delete all content of the folder recursively, see the method here: https://github.com/joomla/joomla-cms/blob/4.2-dev/libraries/src/Filesystem/Folder.php#L272 .

You can try that yourself by creating a temporary folder e.g. inside your Joomla's tmp folder, filling that folder with some subfolders and files, adding a `use Joomla\CMS\Filesystem\Folder;` and a `Folder::delete('/full/path/to/the/previously/mentioned/folder');` to your frontend template's index.php, visit your site and then check if the folder is still there.

**Reply by pjdevries** [reply]
Thanx for the response @richard67. And yes, `Folder::delete` indeed **_does_** delete a folder recursively, including all content. However, the test `if ($files !== false && !\count($files))` prevents  `Folder::delete` from being executed ;)

**Reply by richard67** [reply]
Hmm you are right. But the comment `// Delete any folders that don't have any content in them.` tells what it does, so it seems to be intended.

There is some code before that piece of code which seems to make sure that all files inside the folders will be deleted:

https://github.com/joomla/joomla-cms/blob/4.2-dev/libraries/src/Installer/Adapter/FileAdapter.php#L257-L268

So after that, all folders should be empty, and so the code mentioned by you should be sufficient.

**Reply by pjdevries** [reply]
> There is some code before that piece of code which seems to make sure that all files inside the folders will be deleted:

Not exacttly, as far as I can see. What it does, is delete all files from the top folder, collecting sub folder names along the way. It then processes those collected sub folders, checks if they're empty and leaves them as they are if they are not. The files within those folders are never deleted, so neither will the folders themselves.

**Reply by pjdevries** [reply]
By the way, I'm quite sure because my files do get installed in a folder of my choosing,  but never get uninstalled, except for a file in the top most folder :)

**Reply by richard67** [reply]
> Not exacttly, as far as I can see. What it does, is delete all files from the top folder, collecting folder names along the way.

Not right. It goes through all the file elements in your manifest XML!!! If they refer to a folder, the folder is put into the list, and if they refer to a file, the file is deleted. Then the folders are deleted in the 2nd loop. So if your manifest correctly lists all files and folders, all should be deleted, or not?

**Reply by pjdevries** [reply]
If my manifest didn't correctly list all files and folders, they would not be installed, I guess. I can not be 100% sure though, because it's difficult to figure out how a `file` manifest should look, since I can not find any docs for it. But like I mentioned, all files and folders are installed correctly, but, apart from the one file in the top folder, don't get deleted when uninstalling. 

When stepping through the code with the debugger, I see it happening just as I described. So I don't know what else to say.

**Reply by pjdevries** [reply]
Maybe you want tor try it yourself @richard67. To make that easier, I created a type `file` manifest file to illustrate the issue: [files_manifest_test.zip](https://github.com/joomla/joomla-cms/files/9276915/files_manifest_test.zip). It installs the following files and folders:

```
layouts/files_manifest_test/
layouts/files_manifest_test/foo.php
layouts/files_manifest_test/folder1/
layouts/files_manifest_test/folder1/bar.php
layouts/files_manifest_test/folder2/
layouts/files_manifest_test/folder2/baz.php
```

When uninstalling, It only deletes `layouts/files_manifest_test/foo.php` but leaves everything else.

**Reply by richard67** [reply]
Your files section looks as follows:

```
        <files folder="layouts/files_manifest_test" target="layouts/files_manifest_test">
            <folder>folder1</folder>
            <folder>folder2</folder>
            <file>foo.php</file>
        </files>
```

The code expects it to contain all files so it looks as follows:
```
        <files folder="layouts/files_manifest_test" target="layouts/files_manifest_test">
            <folder>folder1</folder>
            <folder>folder2</folder>
            <file>folder1/bar.php</file>
            <file>folder2/baz.php</file>
            <file>foo.php</file>
        </files>
```

I only want to explain how it currently works.

If that is right or wrong I don't know.

Looking at the Joomla files manifest, it seems indeed to wrong because that does not list all files in the subfolders: https://github.com/joomla/joomla-cms/blob/4.2-dev/administrator/manifests/files/joomla.xml . But that might be because the Joomla files cannot really be uninstalled so nobody ever cared about that.

It would be worth a check how it works with other extension types. I've checked the code in other adapters, and each one handles it differently.

Anyway, if you are sure it is an issue, you should open an issue, because in discussions it might not get the right attendance.

**Reply by pjdevries** [reply]
Thanx for the effort @richard67. I did not consider the possibility of including each individual file and folder in the manifest. Like I mentioned I could not find docs about `file` type manifests and file sections in other manifest types are handled differently, just like you say. 

It's not really up to me to say whether it is an issue or not. I definitely think it is very undesirable that file sections are handled differently across different manifest types. It's also far from ideal having to specify each individual file and folder in a manifest file, just because there is this difference in the deletion logic. But I guess you are right that opening an issue is the only way to give it proper attention, although I'm not well enough equipped to defend my point and counter any objections.  And since it would probably introduce a backward compatibility break, I can already sense more or less what's coming. But I'll give it a try.

**Reply by brianteeman** [reply]
> The code expects it to contain all files so it looks as follows:

That may be how the code works (not checked) but it is not the intended behaviour. Or at least its not what we show people by example.

https://github.com/joomla/joomla-cms/blob/7cf5e1228db337881d7678ad1c71915d5c885150/templates/cassiopeia/templateDetails.xml#L11-L25

Nor is it what the [documentation ](https://docs.joomla.org/Manifest_files)says 

> Files to copy to the front-end directory should be placed in the <files> element. You can use the optional folder attribute to specify a directory in the ZIP package to copy from. Each file to copy must be represented by a <filename> element. If you want to copy an entire folder at once, you can define it as a <folder>.

> Files to copy to the back-end directory should be placed in the <files> element under the <administration>. You can use the optional folder attribute to specify a directory in the ZIP package to copy from. See Front-end files for further rules.

**Reply by richard67** [reply]
@brianteeman You are right, the documentation clearly tells it. So we have at least an issue with the files adapter. Maybe others should be checked, too.

**Reply by pjdevries** [reply]
Thanx for pitching in @brianteeman. Maybe you or @richard67 are better qualified than me to create a proper issue.

**Reply by richard67** [reply]
I will check first how it is with other extension types. I will use the weblinks 4.0.1 package which contains a component, a module and a few plugins, which all in their XML specify folders but not the files inside these folders, like it is written in the documentation. If these delete all files in these folders on unistalling, then it should be done the same way for extensions of type "files". Will make then maybe an issue or maybe already a PR.

So or so I slowly start to understand why I often saw people reporting issues in forums and on social media about leftover files from uninstalled extensions. 😄

**Reply by pjdevries** [reply]
To be honest, I'm not entirely sure that the behaviour of the `FileAdapter` is responsible for all those leftover files. There are also quite a few other sloppinesses :grinning: But solving this will certainly prevent some of them.

**Reply by pjdevries** [reply]
Thanx again for your help.

**Reply by richard67** [reply]
Well all other extension types delete all stuff when uninstalling. Only the files adapter not.

**Reply by richard67** [reply]
Hmm, the weblinks extensions all store their stuff inside their folder, which is deleted anyway at the end, while your files package example stores things in the global layouts folder in a subfolder, e.g. in `JOOMLA_ROOT/layouts/files_manifest_test`.

What if that subfolder already exists and contains files and subfolders e.g. from another extension?

Maybe that was the reason why it was made like it is long time ago?

But if so, then it would have to be documented.

**Reply by richard67** [reply]
The [documentation here](https://docs.joomla.org/Extension_types_(general_definitions)#File) says:

>The File extension type is used to install individual files into Joomla into a custom directory. There are no examples in Joomla Core of this type and it is the least used type however it can be used for example to place CLI scripts into the cli directory within Joomla or to place template overrides into a specific directory.

Beside that, there is no documentation on that type.

But it reads as if you could copy files into existing folders also used by other extensions, especially the example with the template overrides.

And then it could really be desired but undocumented behaviour.

**Reply by pjdevries** [reply]
I see what you mean. Looking at it that way, the current implementation makes sense. However, if it is meant to prevent deletion of files and folders used by other extensions, one may assume it should not be possible to overwrite existing files either. But since a `method="upgrade"` is supposed to overwrite existing files, this seems to pose somewhat of a dilemma.

**Reply by brianteeman** [reply]
The difference would be that to "upgrade" someone elses file you woud need to know it existed.

**Reply by pjdevries** [reply]
That is true. So being able to create a package that can arbitrarily, consciously or unconsciously, overwrite existing files is not entirely desirable, don't you think?

**Reply by brianteeman** [reply]
thats the point - it can't unless you know the exact folder and filename

**Reply by pjdevries** [reply]
I think I'm missing the point you're making. One could do it accidentally. Like in my test example. If files with the same names already exist in the same folders, one would unknowingly overwrite them. I can not imagine that being a good thing.

**Reply by richard67** [reply]
The thing is that the "File" adapter - in opposite to all the others - uses the attribute "target" of the "files" element of the manifest XML, and that allows to copy the files at any place inside the Joomla root on installation.

This is done because an extension of type "file" doesn't have an own extension folder, I assume.

Now nobody can keep an extension developer from doing following in the manifest:
```
    <fileset>
        <files folder="layouts/files_manifest_test" target="layouts/joomla">
            <folder>form</folder>
            <folder>html</folder>
            <file>foo.php</file>
        </files>
    </fileset>
```

Or maybe a less weird, more practical example:
```
    <fileset>
        <files folder="overrides" target="templates/cassiopeia">
            <folder>html</folder>
        </files>
    </fileset>
```

As the installer does not save anywhere which files it has copied from a source folder to the target, we do not know which of the files in the subfolder come from the extension and which files were already there.

And so we cannot really remove the folder and all its contents.

Even if we make the installer to refuse that on installation, there still might be extensions in the wild where this has worked on installation.

**Reply by pjdevries** [reply]
So it appears extension devs need to be extra careful when installing files in this kind of location and always put them in a custom sub folder, with a name not likely to conflict with an existing folder. For a proper uninstall, there seems to be no other way than explicitly specifying each and every file and folder, just as you explained before.

**Reply by richard67** [reply]
Yes, and it possibly needs to update docs by some hints not to use the "folder" element for extension type "file". E.g. in case of your testing extension the files section should look as follows:

```
    <fileset>
        <files folder="layouts/files_manifest_test" target="layouts/files_manifest_test">
            <file>folder1/bar.php</file>
            <file>folder2/baz.php</file>
            <file>foo.php</file>
        </files>
    </fileset>
```

**Reply by pjdevries** [reply]
Indeed, as it would prevent accidental removal of existing folders. 

About those docs. Are there any docs for the `file` type manifest?

**Reply by richard67** [reply]
If you do it like I've written in my previous comment, you can safely install your files into existing folders which are also used by the core or by other 3rd party extensions, and the files would be properly removed on uninstall.

**Reply by pjdevries** [reply]
I will make the necessary changes in my manifest :)

**Reply by richard67** [reply]
>  Are there any docs for the `file type` manifest?

Not that I knew of, except of the one I had linked before which only briefly explains what that type is good for: https://docs.joomla.org/Extension_types_(general_definitions)#File

In the documentation for manifest files this type is not described, also not the "target" attribute of the "files" element: https://docs.joomla.org/Manifest_files

**Reply by richard67** [reply]
> I will make the necessary changes in my manifest :)

Let us know if it works because I haven't tested my suggestion yet.

**Reply by pjdevries** [reply]
I'll do that later today and let you know.

**Comment by pjdevries** [comment]
Apologies for not keeping my "later today" promise. I just tested a few variations, starting with the one you suggested last:

```xml
<fileset>
    <files folder="layouts/files_manifest_test" target="layouts/files_manifest_test">
        <file>folder1/bar.php</file>
        <file>folder2/baz.php</file>
        <file>foo.php</file>
    </files>
</fileset>
```

This variation installs all files and folders properly. However, on uninstallation, it deletes all files and folders, except the top folder `files_manifest_test`. It can be remedied by explicitly including that folder in the files & folders list and excluding it from the `<files>` `folder` and `target` attributes:

```xml
<fileset>
    <files folder="layouts" target="layouts">
        <folder>files_manifest_test</folder>
        <file>files_manifest_test/folder1/bar.php</file>
        <file>files_manifest_test/folder2/baz.php</file>
        <file>files_manifest_test/foo.php</file>
    </files>
</fileset>
```
With this arrangement all files and folders are properly installed and deleted on uninstallation.
